### ISSUE TITLE
switch_order_layer() got an unexpected keyword argument 'reshape'

### ISSUE THREAD

**Issue by huayong** [PaddlePaddle/Paddle#5877]
问题如title，最新在mac上编译的代码，commit为659c937

**Comment by QiJune** [comment]
请问是编译错误吗？还是使用错误
如果是使用错误，能贴一下网络配置吗？

**Comment by huayong** [comment]
是使用错误，配置如下：
import paddle.v2 as paddle
import numpy as np
#LAYER_TYPE = 'cudnn_conv'
LAYER_TYPE = 'exconv'
def conv_relu(input, nout,name=None, ks=3, stride=1,
        pad=1, num_channels=None, dropout_ratio=0,
        param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005),
        bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0)):
    conv = paddle.layer.img_conv(input, name=name, filter_size=ks, stride=stride,
            num_filters=nout, padding=pad, act=paddle.activation.Relu(), num_channels=num_channels,
            layer_attr=paddle.attr.Extra(drop_rate=dropout_ratio),
            param_attr=param_attr,
            bias_attr=bias_attr,
            layer_type=LAYER_TYPE)
    return conv

def max_pool(input, ks=2, stride=2):
    return paddle.layer.img_pool(input=input, stride=stride, pool_size=ks)

def fcn(input, label):

    # the base net
    conv1_1 = conv_relu(input, 64, pad=100, num_channels=3)
    conv1_2 = conv_relu(conv1_1, 64)
    pool1 = max_pool(conv1_2)

    conv2_1 = conv_relu(pool1, 128)
    conv2_2 = conv_relu(conv2_1, 128)
    pool2 = max_pool(conv2_2)

    conv3_1 = conv_relu(pool2, 256)
    conv3_2 = conv_relu(conv3_1, 256)
    conv3_3 = conv_relu(conv3_2, 256)
    pool3 = max_pool(conv3_3)

    conv4_1 = conv_relu(pool3, 512)
    conv4_2 = conv_relu(conv4_1, 512)
    conv4_3 = conv_relu(conv4_2, 512)
    pool4 = max_pool(conv4_3)

    conv5_1 = conv_relu(pool4, 512)
    conv5_2 = conv_relu(conv5_1, 512)
    conv5_3 = conv_relu(conv5_2, 512)
    pool5 = max_pool(conv5_3)

    # fully conv
    fc6 = conv_relu(pool5, 4096, name='__fc_layer_0__', ks=7, pad=0, dropout_ratio=0)

    fc7 = conv_relu(fc6, 4096, name='__fc_layer_1__', ks=1, pad=0, dropout_ratio=0)

    score_fr = paddle.layer.img_conv(fc7, num_filters=21,
                                    filter_size=1,
                                    padding=0,
                                    layer_type=LAYER_TYPE,
                                    param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005, initial_mean=1, initial_std=0),
                                    bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0, initial_mean=1, initial_std=0)
                                    )
    upscore = paddle.layer.img_conv(score_fr,
                   num_filters=21,
                   filter_size=64,
                   trans=True,
                   stride=32,
                   layer_type="exconvt",
                   bias_attr=False,
                   param_attr=paddle.attr.Param(learning_rate=0, l2_rate=0)
                   )
    score = paddle.layer.crop(name="crop_0", input=[upscore, input], axis=2, offset=[19,19],)
    softmax = paddle.layer.switch_order(score,
                                        reshape={'height':[0,1,2], 'width':[3]},
                                        act=paddle.activation.Softmax())
    loss = paddle.layer.cross_entropy_cost(input=softmax,
                        label=label)
    return loss, softmax




def upsample_filt(size):
    """
    Make a 2D bilinear kernel suitable for upsampling of the given (h, w) size.
    """
    factor = (size + 1) // 2
    if size % 2 == 1:
        center = factor - 1
    else:
        center = factor - 0.5
    og = np.ogrid[:size, :size]
    return (1 - abs(og[0] - center) / factor) * \
           (1 - abs(og[1] - center) / factor)

def interp(m, k, h, w):
    """
    Set weights of each layer in layers to bilinear kernels for interpolation.
    """
    data = np.zeros((m, k, h, w))
    if m != k and k != 1:
        print 'input + output channels need to be the same or |output| == 1'
        raise
    if h != w:
        print 'filters need to be square'
        raise
    filt = upsample_filt(h)
    data[range(m), range(k), :, :] = filt
    return data

if __name__ == '__main__':
    input = paddle.layer.data(
        name="image", type=paddle.data_type.dense_vector(3*224*224))
    label = paddle.layer.data(
        name="label", type=paddle.data_type.dense_vector(224*224))
    fcn(input, label)

**Comment by huayong** [comment]
是使用错误，配置如下：
`import paddle.v2 as paddle
import numpy as np
#LAYER_TYPE = 'cudnn_conv'
LAYER_TYPE = 'exconv'
def conv_relu(input, nout,name=None, ks=3, stride=1,
        pad=1, num_channels=None, dropout_ratio=0,
        param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005),
        bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0)):
    conv = paddle.layer.img_conv(input, name=name, filter_size=ks, stride=stride,
            num_filters=nout, padding=pad, act=paddle.activation.Relu(), num_channels=num_channels,
            layer_attr=paddle.attr.Extra(drop_rate=dropout_ratio),
            param_attr=param_attr,
            bias_attr=bias_attr,
            layer_type=LAYER_TYPE)
    return conv

def max_pool(input, ks=2, stride=2):
    return paddle.layer.img_pool(input=input, stride=stride, pool_size=ks)

def fcn(input, label):

    # the base net
    conv1_1 = conv_relu(input, 64, pad=100, num_channels=3)
    conv1_2 = conv_relu(conv1_1, 64)
    pool1 = max_pool(conv1_2)

    conv2_1 = conv_relu(pool1, 128)
    conv2_2 = conv_relu(conv2_1, 128)
    pool2 = max_pool(conv2_2)

    conv3_1 = conv_relu(pool2, 256)
    conv3_2 = conv_relu(conv3_1, 256)
    conv3_3 = conv_relu(conv3_2, 256)
    pool3 = max_pool(conv3_3)

    conv4_1 = conv_relu(pool3, 512)
    conv4_2 = conv_relu(conv4_1, 512)
    conv4_3 = conv_relu(conv4_2, 512)
    pool4 = max_pool(conv4_3)

    conv5_1 = conv_relu(pool4, 512)
    conv5_2 = conv_relu(conv5_1, 512)
    conv5_3 = conv_relu(conv5_2, 512)
    pool5 = max_pool(conv5_3)

    # fully conv
    fc6 = conv_relu(pool5, 4096, name='__fc_layer_0__', ks=7, pad=0, dropout_ratio=0)

    fc7 = conv_relu(fc6, 4096, name='__fc_layer_1__', ks=1, pad=0, dropout_ratio=0)

    score_fr = paddle.layer.img_conv(fc7, num_filters=21,
                                    filter_size=1,
                                    padding=0,
                                    layer_type=LAYER_TYPE,
                                    param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005, initial_mean=1, initial_std=0),
                                    bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0, initial_mean=1, initial_std=0)
                                    )
    upscore = paddle.layer.img_conv(score_fr,
                   num_filters=21,
                   filter_size=64,
                   trans=True,
                   stride=32,
                   layer_type="exconvt",
                   bias_attr=False,
                   param_attr=paddle.attr.Param(learning_rate=0, l2_rate=0)
                   )
    score = paddle.layer.crop(name="crop_0", input=[upscore, input], axis=2, offset=[19,19],)
    softmax = paddle.layer.switch_order(score,
                                        reshape={'height':[0,1,2], 'width':[3]},
                                        act=paddle.activation.Softmax())
    loss = paddle.layer.cross_entropy_cost(input=softmax,
                        label=label)
    return loss, softmax




def upsample_filt(size):
    """
    Make a 2D bilinear kernel suitable for upsampling of the given (h, w) size.
    """
    factor = (size + 1) // 2
    if size % 2 == 1:
        center = factor - 1
    else:
        center = factor - 0.5
    og = np.ogrid[:size, :size]
    return (1 - abs(og[0] - center) / factor) * \
           (1 - abs(og[1] - center) / factor)

def interp(m, k, h, w):
    """
    Set weights of each layer in layers to bilinear kernels for interpolation.
    """
    data = np.zeros((m, k, h, w))
    if m != k and k != 1:
        print 'input + output channels need to be the same or |output| == 1'
        raise
    if h != w:
        print 'filters need to be square'
        raise
    filt = upsample_filt(h)
    data[range(m), range(k), :, :] = filt
    return data

if __name__ == '__main__':
    input = paddle.layer.data(
        name="image", type=paddle.data_type.dense_vector(3*224*224))
    label = paddle.layer.data(
        name="label", type=paddle.data_type.dense_vector(224*224))
    fcn(input, label)`

**Comment by huayong** [comment]
是使用错误，配置如下，fcn网络：
`
def conv_relu(input, nout,name=None, ks=3, stride=1,
        pad=1, num_channels=None, dropout_ratio=0,
        param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005),
        bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0)):
    conv = paddle.layer.img_conv(input, name=name, filter_size=ks, stride=stride,
            num_filters=nout, padding=pad, act=paddle.activation.Relu(), num_channels=num_channels,
            layer_attr=paddle.attr.Extra(drop_rate=dropout_ratio),
            param_attr=param_attr,
            bias_attr=bias_attr,
            layer_type=LAYER_TYPE)
    return conv

def max_pool(input, ks=2, stride=2):
    return paddle.layer.img_pool(input=input, stride=stride, pool_size=ks)

def fcn(input, label):

    # the base net
    conv1_1 = conv_relu(input, 64, pad=100, num_channels=3)
    conv1_2 = conv_relu(conv1_1, 64)
    pool1 = max_pool(conv1_2)

    conv2_1 = conv_relu(pool1, 128)
    conv2_2 = conv_relu(conv2_1, 128)
    pool2 = max_pool(conv2_2)

    conv3_1 = conv_relu(pool2, 256)
    conv3_2 = conv_relu(conv3_1, 256)
    conv3_3 = conv_relu(conv3_2, 256)
    pool3 = max_pool(conv3_3)

    conv4_1 = conv_relu(pool3, 512)
    conv4_2 = conv_relu(conv4_1, 512)
    conv4_3 = conv_relu(conv4_2, 512)
    pool4 = max_pool(conv4_3)

    conv5_1 = conv_relu(pool4, 512)
    conv5_2 = conv_relu(conv5_1, 512)
    conv5_3 = conv_relu(conv5_2, 512)
    pool5 = max_pool(conv5_3)

    # fully conv
    fc6 = conv_relu(pool5, 4096, name='__fc_layer_0__', ks=7, pad=0, dropout_ratio=0)

    fc7 = conv_relu(fc6, 4096, name='__fc_layer_1__', ks=1, pad=0, dropout_ratio=0)

    score_fr = paddle.layer.img_conv(fc7, num_filters=21,
                                    filter_size=1,
                                    padding=0,
                                    layer_type=LAYER_TYPE,
                                    param_attr=paddle.attr.Param(learning_rate=1e-10, l2_rate=0.0005, initial_mean=1, initial_std=0),
                                    bias_attr=paddle.attr.Param(learning_rate=2e-10, l2_rate=0, initial_mean=1, initial_std=0)
                                    )
    upscore = paddle.layer.img_conv(score_fr,
                   num_filters=21,
                   filter_size=64,
                   trans=True,
                   stride=32,
                   layer_type="exconvt",
                   bias_attr=False,
                   param_attr=paddle.attr.Param(learning_rate=0, l2_rate=0)
                   )
    score = paddle.layer.crop(name="crop_0", input=[upscore, input], axis=2, offset=[19,19],)
    softmax = paddle.layer.switch_order(score,
                                        reshape={'height':[0,1,2], 'width':[3]},
                                        act=paddle.activation.Softmax())
    loss = paddle.layer.cross_entropy_cost(input=softmax,
                        label=label)
    return loss, softmax

if __name__ == '__main__':
    input = paddle.layer.data(
        name="image", type=paddle.data_type.dense_vector(3*227*227))
    label = paddle.layer.data(
        name="label", type=paddle.data_type.dense_vector(227*227))
    fcn(input, label)
`

**Comment by wanghaoshuang** [comment]
@huayong
这个op后来有人改动过,请参考如下doc： 
https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/layers.py#L6809

**Comment by huayong** [comment]
把switch_order的reshape属性改成reshape_axis=3，问题已经解决

**Comment by huayong** [comment]
@wanghaoshuang
虽然网络可以正常运行，但是switch_order op后得到的softmax的size、width和height都是0，其他layer都有正常的值
在自己编译的版本和直接pip install的都有这个问题？

**Comment by wanghaoshuang** [comment]
我们在python端没有推算output的size、height和width,
因为fcn net 的input的size、height和width都是变化的，所以在构建网络的时候是无法正确推算每一层大小的。
只有在运行时，我们会在c++代码里根据实际数据推算每层大小。

**Comment by wanghaoshuang** [comment]
>switch_order op后得到的softmax的size、width和height都是0
@huayong  你怎么获得的size、width和height？

**Comment by huayong** [comment]
就是上面的代码，debug模式下看到的，前面的layer在运行完都正常，可以看到正确的width、height和size，输出也正常，但是运行完switch_order后看到的相关数据都是0

**Comment by wanghaoshuang** [comment]
@huayong 贴一下debug log？

**Comment by qingqing01** [comment]
Python里打印`switch_order`之后的height, width是0，应该没问题。[`switch_order`](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer/config_parser.py#L3847)这里没有设置`height`,`width`大小。对于size也是同样的问题，这里没有设置。但在C++代码里，如@wanghaoshuang所说，会动[态获取每个mini-batch输入数据的大小](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/SwitchOrderLayer.cpp#L84)。

**Comment by wanghaoshuang** [comment]
@huayong 我明白问题所在了，我稍后修改下。

**Comment by peterzhang2029** [comment]
Closing due to low activity. Feel free to reopen it.

**Comment by chengyuz** [comment]
Hi，我这边输入数据：
softmax = paddle.layer.switch_order(
        name = 'softmax',
        input = upscore,
        reshape_axis = 3,
        act = paddle.activation.Softmax())
对应标签：
label = paddle.layer.data(
            name = 'label',
            type = paddle.data_type.dense_vector(cfg.IMG_HEIGHT * cfg.IMG_WIDTH))
其中softmax输出大小是N*H*W*C，损失函数：
loss = paddle.layer.cross_entropy_cost(input = softmax, label = label)
不过训练过程中报错：
F0528 10:15:50.201407 24479 Matrix.cpp:851] Check failed: output_ptr && label_ptr Invalid argument pointer
*** Check failure stack trace: ***
    @     0x7fbae0add06d  google::LogMessage::Fail()
    @     0x7fbae0ae0b1c  google::LogMessage::SendToLog()
    @     0x7fbae0adcb93  google::LogMessage::Flush()
    @     0x7fbae0ae202e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fbae08e7202  paddle::GpuMatrix::oneHotCrossEntropy()
    @     0x7fbae0752345  paddle::CostLayer::forward()
    @     0x7fbae06d5999  paddle::NeuralNetwork::forward()
    @     0x7fbae06d6633  paddle::GradientMachine::forwardBackward()
    @     0x7fbae0ab9539  GradientMachine::forwardBackward()
    @     0x7fbae068339c  _wrap_GradientMachine_forwardBackward
    @     0x7fbaffd58990  PyEval_EvalFrameEx
    @     0x7fbaffd5b4e9  PyEval_EvalCodeEx
    @     0x7fbaffd58482  PyEval_EvalFrameEx
    @     0x7fbaffd5b4e9  PyEval_EvalCodeEx
    @     0x7fbaffd58482  PyEval_EvalFrameEx
    @     0x7fbaffd5b4e9  PyEval_EvalCodeEx
    @     0x7fbaffd58482  PyEval_EvalFrameEx
    @     0x7fbaffd5b4e9  PyEval_EvalCodeEx
    @     0x7fbaffd58482  PyEval_EvalFrameEx
    @     0x7fbaffd5b4e9  PyEval_EvalCodeEx
    @     0x7fbaffd5b70a  PyEval_EvalCode
    @     0x7fbaffd7493d  run_mod
    @     0x7fbaffd75ab8  PyRun_FileExFlags
    @     0x7fbaffd76cd8  PyRun_SimpleFileExFlags
    @     0x7fbaffd88d3c  Py_Main
    @     0x7fbafefa4f45  __libc_start_main
    @     0x7fbb0027987f  (unknown)
Aborted
